### PR TITLE
Enables linting across the entire project

### PR DESCRIPTION
**Pull request by schuylermartin45**
- Enables lint checking in `pre-commit`
- Skips `examples/` directory
- Fixes type annotation with `Variant`; Charles confirmed this change
- Ensures all files (non-examples) have `annotations` from `__future__`

### CODE REVIEW

**Comment by schuylermartin45** [comment]
This is a pre-curser to considering to enable the static analyzer. Approximately 250 analyzer errors (out of nearly 3000) got cleaned up as a consequence of making the linter happy.

File: percy/repodata/repodata.py [hunk 1]
```diff
@@ -122,8 +124,6 @@ def compare_package_with_defaults(package: Package, defaults_pkgs: dict[str, dic
                     "defaults_version": defaults_version,
                     "defaults_build_number": defaults_build_number,
                 }
-    except InvalidVersionSpec as exc:
-        logging.error(f"Exception: {local_name} {exc}")
-    except ValueError as exc:
-        logging.error(f"Exception: {local_name} {exc}")
+    except (ValueError, InvalidVersionSpec) as exc:
+        logging.error("Exception: %s %s", local_name, exc)
```
Comment (JeanChristopheMorinPerso) [inline percy/repodata/repodata.py:128]: Tip: `logging.exception("local_name")` will take care of logging the exception stack.

**Review by JeanChristopheMorinPerso — APPROVED** [review]
From the look of it, everythuing seems good to me. There is inconsistencies in the docstring style used (some are numpy/google others are sphinx) and there is other inconsistencies, but that was already the case before.

**Comment by schuylermartin45** [comment]
The linter will accept both types of docstrings and I haven't decided on one over the other yet. We should discuss as a team/group and pick one.

I would lean towards Sphinx so we can auto-generate docs and I find it is a little easier to work with and probably easier for a computer to parse.

One thing we definitely shouldn't do is duplicate the typing info in the docstrings as most tools should be able to parse both and `mypy`/VS Code only looks at the type annotations.

**Pull request merged**